### ISSUE TITLE
Support for c# 7 value tuple destructuring

### ISSUE THREAD

**Issue by jotabe-net** [serilog/serilog#965]
If you try to destructure a C# 7 Value Tuple, by using `{@tupleValue}` it doesn't destructure it.

For example, when this code is run
```
var x =1;
var y = 7;
var point = (x,y);

Log.Information("The point is {@point}", point);
```

You get this message logged:
```The point is ValueTuple`2```
and this property:
```point   {_typeTag: "ValueTuple`2"}```

I tried to implement an `IDestructuringPolicy`, but it wasn't even invoked. I suppose that's because it's a struct (value type) instead of an object (reference type)

However I think it would be interesting to support this destructuring as tuples will become widely used sooner or later.

These are some clues for the possible implementation:
- it's a value type instead of a reference type
- it has fields, instead of properties
- the fileds have the names: `Item1`, `Item2` and so on
- it is supposed to automatically receive a `TupleElementNamesAttribute` with the friendly names of the fileds *in a pre-order depth-first traversal of a type's construction*, as stated in [this atrribute 's documentation](https://msdn.microsoft.com/en-us/library/system.runtime.compilerservices.tupleelementnamesattribute.tupleelementnamesattribute(v=vs.110).aspx#Anchor_1) I tried to check this in the debugger, but the debugger is failing when dealing with this kind of type and Reflection.

In the meantime, you can solve this problem by manually converting it to an anonymus type, like so:
```
Log.Information("The point is {@point}", new { point.x, point.y});
```

**Comment by nblumhardt** [comment]
Thanks for the note. Yes, this looks like a worthwhile improvement to explore.

I guess there are these implementation alternatives:

 1. Only support it for target frameworks that include `ValueTuple` natively (not sure yet which these are)
 2. Support it everywhere, by depending on the NuGet package carrying the `ValueTuple` type
 3. Support it everywhere, with an add-on package
 4. Support it everywhere, using reflection only (no direct dependency on `ValueTuple` itself)

It seems like a combination of 1 & 3 could work, i.e. use `#if VALUETUPLE` to conditionally support it in Serilog itself, and distribute a NuGet package version that would need to be activated via `Destructure.ValueTuples()` for use on other TFMs.

Anyone keen to investigate?

**Comment by nblumhardt** [comment]
I've been looking into this some more; I think the best representation in Serilog for tuples is, in fact, arrays...

The issue with `ValueTuple<>` is that the member names are only available through methods that consume or return tuples, not the raw tuple objects themselves (there are some discussions online around this already).

Given that we're back to then using `Item0`, `Item1` and so-on, we might as well use `[1, 2]` as the representation.

Since tuples are primitive value types, one shouldn't need to opt in with `@` to serialize them. We already have special handling of arrays under regular capturing, so producing arrays from tuples doesn't seem too far from what we have now.... Any thoughts?

Planning a first cut that targets .NET 4.7 only, deferring the decision about whether or not to take a dependency on the System.ValueTuple (package) on other platforms.

**Comment by nblumhardt** [comment]
This shipped in Serilog 2.5.
